change computation of hash value.

funny-falcon opened the pull request:
To protect against Hash DoS, change the way hash value is computed.
Class|Struct should define method `def hash(hasher)` and call
`hasher << @ivar` inside.

As an option, for speed, and for backward compatibility, `def hash`
still could be implemented. It will be used for Hash of matched type.
`Thread#hash` and `Signal#hash` is implemented as unseeded cause they are
 used before `StdHasher @@seed` is initialized.

But it is better to implement `def hash(hasher)`.

StdHasher is default hasher that uses `hash(hasher)` and it is used as default
seeded hasher. It also implements `unseeded` for `Enums`.

Also, number normalization for hashing introduced, ie rule 'equality
forces hash equality' is forced (`a == b` => `a.hash == b.hash`).
Normalization idea is borrowed from Python implementation.
(idea by Akzhan Abdulin @akzhan)

Fixes #4578
Prerequisite for #4557
Replaces #4581

funny-falcon commented:
Fresh PR in place of #4621

akzhan commented on src/big/big_float.cr:
```diff
@@ -17,6 +17,18 @@ struct BigFloat < Float
     LibGMP.mpf_init_set_str(out @mpf, str, 10)
   end
 
+  def initialize(num : BigInt)
```
This pull request is much broader than simply changing the hashing calculation. But these updates are really important.

akzhan replied:
Latest builds of #4653 shows unstable behavior of constructors (Compiler can prefer another one under some conditions).

Please remove 'em and use `to_big_X`.

akzhan commented on src/float.cr:
```diff
@@ -148,13 +149,33 @@ struct Float32
+      {% if flag?(:x86) || flag?(:x86_64) || flag(:arm) || flag(:aarch64) %}
+	# it should work on every architecture where endianess of Float32 and Int32
+	# matches and float is IEEE754.
+	unsafe_int = unsafe_as(Int32)
```
Is it result of `crystal tool format`? Strange indentation.

konovod replied:
Afaik `tool format` don't touch macro (`{% %}`) lines, so they should be formatted manually.

funny-falcon replied:
I removed usage of macros here in following commits.

straight-shoota commented on src/float.cr:
```diff
@@ -148,13 +149,33 @@ struct Float32
+  def hash_normalize
+    float_normalize_wrap do
+      {% if flag?(:x86) || flag?(:x86_64) || flag(:arm) || flag(:aarch64) %}
+	# it should work on every architecture where endianess of Float32 and Int32
```
This needs formatting.

funny-falcon replied:
Can you hint which way?
I relied on `crystal tool format`, but it doesn't help much.

straight-shoota replied:
Just apply proper indentation on this and the following lines. This line should be indented by 8 spaces for example.

akzhan replied:
@funny-falcon indent by two spaces.

```crystal
  def hash_normalize
    float_normalize_wrap do
      {% if flag?(:x86) || flag?(:x86_64) || flag(:arm) || flag(:aarch64) %}
        # it should work on every architecture where endianess of Float32 and Int32
```

no tabs please.

funny-falcon replied:
I see: i used tabs unintentionally instead of spaces.
Fixed that, and reduced relying on macroses.

straight-shoota commented on src/float.cr:
```diff
@@ -235,28 +235,28 @@ struct Float64
-      {% else %}
-	float_normalize_reference
-      {% end %}
+      if FLOAT64_IS_IEEE754
```
Why is this no longer a macro as in the first commit? Can this constant change at runtime?

sdogruyol replied:
That's not possible

straight-shoota replied:
Sry, I meant if it can differ based on the system where the code is executed or is it fixed for every target architecture? If fixed, the appropriate branch should be chosen according to target arch at compile time.

funny-falcon replied:
Unfortunately, you are right: it is tested in runtime even in release build.
Lets think a bit more.

funny-falcon replied:
looks like crystal compiler does very strange things with constants:
even if it knows constant value at compile time and can use it in macros, it still inserts checks for constant's initializer in final code, and doesn't propagate them as real constants to llvm optimizer.

I add `private def hash_bits` and `private def hash_modulus` to workaround this issue.

akzhan replied:
/cc @asterite

akzhan commented on src/number/hash_normalize.cr:
```diff
@@ -0,0 +1,115 @@
+module Number::HashNormalize
+  # Idea by Akzhan Abdulin @akzhan
```
Please drop this line, thanks.

akzhan commented on src/number/hash_normalize.cr:
```diff
@@ -0,0 +1,115 @@
+  # exponentiation algorithm.  For reduce(2**e) it's even better: since
+  # P is of the form 2**n-1, reduce(2**e) is 2**(e mod n), and multiplication
+  # by 2**(e mod n) modulo 2**n-1 just amounts to a rotation of bits.
+  #
```
And this empty comment too.

akzhan commented:
Now LFTM. Glad to see update on number normalization that's missed in Crystal before.

sdogruyol commented:
This is great 😍

funny-falcon commented:
One question: does anyone uses Crystal on architectures with restricted unaligned read?
Looks like x86, x86_64 are certainly allow unaligned access.
arm and aarch64 allows to do unaligned read under certain conditions, but I'm not sure if program compilled with Crystal falls into that conditions.

Problem line is https://github.com/crystal-lang/crystal/pull/4675/files?diff=unified#diff-b88fe795012d5923ca24e66769201422R158 - permuting slice of bytes.

Can any one with access to arm with crystal run following, please:
```crystal
   str = "1234567"
   buf = str.to_slice
   ptr = buf.to_unsafe
   4.times do |i|
     p (ptr+i).as(Pointer(UInt32)).value.to_s(16)
   end
```

RX14 commented:
@funny-falcon We probably do unaligned reads all over the stdlib. As long as there's a spec which will break and alert us to the problem on any new architectures with this limitation it's fine to do unaligned reads for now.

ysbaddaden commented:
Asterite has a nice writeup about ARM and unaligned accesses:
http://antirez.com/news/111

ARMv5 doesn't allow unaligned accesses. ARMv6 allows word-sized unaligned accesses but fails on multiple words. The Linux kernel will rescue and fix unaligned accesses, so unaligned accesses do work on Linux on ARM, but with slow performance. As stated by @RX14 we probably have unaligned accesses in the core/stdlib, thought that doesn't mean we shouldn't care.

I tried to a qemu VM but the ARM emulator allows unaligned accesses, so I can't check what happens in `/proc/cpu/alignment`.

ysbaddaden commented:
Note that LLVM can be smart and fix unaligned accesses at compile time when it's obvious.

funny-falcon commented:
I've tested equivalent C program on Raspberry Pi2, and it handles unaligned read well, so I think no issue there.

akzhan commented on src/stdhasher.cr:
```diff
@@ -0,0 +1,199 @@
+#   end
+
+# StdHasher used as standard hasher in `Object#hash`
+# It have to provide defenense against HashDos, and be reasonably fast.
```
defenense > defense

funny-falcon replied:
done.

akzhan commented on src/number/hash_normalize.cr:
```diff
@@ -0,0 +1,110 @@
+    int_to_hashnorm 314159
+  end
+
+  # This function is for reference implementation.
```
You should note that this method is used under some conditions.

funny-falcon replied:
I improved comment.
I mentioned, that currently hash for BigFloat is not precise for numbers with big fractional part (ie it doesn't distinguish '1.0000000000000001' and '1.0000000000000002').
It happens, cause `BigFloat#hash_normalize` falls back to `to_f.hash_normalize` for such numbers.
It is possible to improve, using this `float_normalize_reference` and overloading `Math.frexp`. (or introducing `BigFloat#frexp` and using it).

Should it be done? (ie, is hashing of such numbers is common task)?
If yes, should it be done in this PR?

akzhan replied:
Feel free to merge #4653 :-)

akzhan replied:
But no, I think that's hashing of Big integers/floats is very uncommon task.

funny-falcon replied:
Those implementation doesn' help, cause they still returns {Float64, Int}, so they strictly equal to `Math.frexp v.to_f`

funny-falcon replied:
Implementation suitable for exact hashing should look like:
```crystal
module Math
  def frexp(value : BigFloat)
    LibGMP.mpf_get_d_2exp(out exp, value)
    frac = BigFloat.new { |mpf|
      if exp >= 0      
        LibGMP.mpf_div_2exp(mpf, self, exp)
      else
        LibGMP.mpf_mul_2exp(mpf, self, -exp)
      end
    }
    {frac, exp}
  end
end
```

funny-falcon replied:
> But no, I think that's hashing of Big integers/floats is very uncommon task.

It is good.
Note, that BigInteger still hashed exactly. Also, BigFloat with big integer part and small fractional part are also hashed exactly.
Stop... I think, my last sentence is not true. 61bit Mersen Prime is larger than Float64 precision (53bit), so precision will be lost with conversion `to_f.hash_normalize` :-(
I should think about carefully.

funny-falcon replied:
looks like, exact hashing for BigFloat is a single reliable way.

akzhan commented on src/number/hash_normalize.cr:
```diff
@@ -73,6 +73,14 @@ module Number::HashNormalize
+  # BigRational hash calculation for values with big fractional part.
+  # (currently it is correct in a meaning "a==b => a.hash==b.hash", but
+  #  1.0000000000000001.hash == 1.0000000000000002.hash).
+  # To do that, Math.frexp should be overloaded for BigFloat and BigRational.
```
I think that frexp has no sense for rationals (only for parts).

akzhan replied:
Was wrong with it, think about complex :)

Sija commented on src/big/big_rational.cr:
```diff
@@ -41,6 +41,22 @@ struct BigRational < Number
 
+  # Creates a exact representation of float as rational.
+  #
+  # It sures that `BigRational.new(f) == f`
```
sures -> ensures

Sija commented on src/big/big_int.cr:
```diff
@@ -267,8 +267,10 @@ struct BigInt < Int
-  def hash
-    to_u64
+  def hash_normalize
+    # remainder(hash_modulus)
```
debug leftover?

funny-falcon replied:
No, implementation description.

Sija commented on src/big/big_rational.cr:
```diff
@@ -139,8 +159,14 @@ struct BigRational < Number
-  def hash
-    to_f64.hash
+  def hash_normalize
+    # self.remainder(hash_modulus).to_f.hash_normalize
```
ditto

funny-falcon replied:
It is implementation description

Sija commented on src/hash.cr:
````diff
@@ -710,14 +710,19 @@ class Hash(K, V)
   # ```
   # foo = {"foo" => "bar"}
-  # foo.hash # => 3247054
+  # foo.hash # => 3247054 (not exactly)
````
not exactly -> approximation

funny-falcon replied:
it is "example". In fact, will be different on every process run.

Sija commented on src/stdhasher.cr:
```diff
@@ -0,0 +1,197 @@
+require "crystal/system/random"
+
+# Hasher usable for `def hash(hasher)` should satisfy protocol:
+#   class MyHasher
```
Please, use triple backticks (```) for code blocks

Sija commented on src/stdhasher.cr:
```diff
@@ -0,0 +1,197 @@
+    high = (v >> 32).to_u32
+    # This condition here cause of some 32bit issue in LLVM binding,
+    # so compiler_spec doesn't pass without it.
+    # Fill free to comment and debug.
```
fill -> feel

Sija commented on src/stdhasher.cr:
```diff
@@ -0,0 +1,197 @@
+# It have to provide defense against HashDos, and be reasonably fast.
+# To protect against HashDos, it is seeded with secure random, and have
+# permutation that hard to forge without knowing seed and seeing hash digest.
+struct StdHasher
```
Why not just `Hasher`?

funny-falcon replied:
Someone suggested, I don't remember exactly who. Why not `StdHasher`?
There is obviously could be different implementations, still usable with suggested protocol, so it is just 'standard hasher'.
But if you really think it should be `Hasher`, I'll rename it.

Sija replied:
I get your drift, yet this `Std` abbreviation looks IMO odd, it would be a precedent in stdlib. Maybe someone have a better name?

akzhan replied:
`Hash::Hasher`?

funny-falcon replied:
I renamed to `Hash::Hasher`.

akzhan commented:
Btw, you can exec bin/crystal docs your files to preview generated documentation. Or make doc afair

funny-falcon commented:
@akzhan 
```
$ bin/crystal doc
Using compiled compiler at `.build/crystal'
Error in line 1: while requiring "./src/**"

in src/crystal/system/unix/arc4random.cr:1: expanding macro

{% skip_file unless flag?(:openbsd) %}
^

in src/crystal/system/unix/arc4random.cr:1: undefined macro variable 'skip_file'

{% skip_file unless flag?(:openbsd) %}
   ^~~~~~~~~
```

Sija commented on src/hash.cr:
````diff
@@ -706,12 +706,8 @@ class Hash(K, V)
-  # foo.hash # => 3247054 (not exactly)
-  # ```
+  # Protocol method for generic hashing.
+  # Note: it should be independent of iteration order.
````
Note -> NOTE

funny-falcon replied:
done

RX14 commented:
@funny-falcon That means your compiler version is too old.

funny-falcon commented:
@RX14 how could be `bin/crystal` too old? I thought it runs just compiled binary.

RX14 commented:
@funny-falcon maybe you haven't recompiled the binary in too long. Either way you somehow are using a compiler from before `skip_file` was introduced.

funny-falcon commented:
@RX14 then you can say, where it is defined.
github doesn't know where it is: https://github.com/crystal-lang/crystal/search?utf8=%E2%9C%93&q=skip_file&type=

funny-falcon commented:
@RX14 it was renamed to "skip": https://github.com/crystal-lang/crystal/blob/master/src/compiler/crystal/macros/methods.cr#L22

akzhan commented:
Didn't found this macro too.

```
➜  crystal git:(master) ✗ grep -r skip_file .
./src/crystal/system/unix/arc4random.cr:{% skip_file unless flag?(:openbsd) %}
./src/crystal/system/unix/getrandom.cr:{% skip_file unless flag?(:linux) %}
./src/crystal/system/unix/sysconf_cpucount.cr:{% skip_file if flag?(:openbsd) || flag?(:freebsd) %}
./src/crystal/system/unix/sysctl_cpucount.cr:{% skip_file unless flag?(:openbsd) || flag?(:freebsd) %}
./src/crystal/system/unix/urandom.cr:{% skip_file if flag?(:openbsd) && flag?(:linux) %}
```

funny-falcon commented:
https://github.com/crystal-lang/crystal/commit/c01a6c7397946838a74e98ed6ec5f72e76af0158
but then used as 'skip_file'
https://github.com/crystal-lang/crystal/commit/971e4fb37cc02ac46e5fa187b9214f2ea90235e8

RX14 commented:
Ok, i'm very confused. How did this ever compile then?

akzhan commented:
@funny-falcon it's kind of missed replace I suppose. But it doesn't fail the build because these platforms aren't compiled on CI I suppose.

RX14 commented:
@akzhan but every platform uses `skip_file`

funny-falcon commented:
Those files aren't included when compiler is building.
Only documentation builder sees them.

funny-falcon commented:
Correction: when file is required on binary build and it should not be skipped, then 'skip_file' is not called cause if 'flag?' condition.
But documentation builder tries to call it because it analyzes every file.

Files are not required on binary build cause if this, for example: https://github.com/crystal-lang/crystal/blob/11e76e49f2f48f370c9e82c48c59cd8343e25838/src/crystal/system/random.cr#L21

kostya commented:
little compared performance,

on master
`./bin/crystal eval 't = Time.now; s = 0_i64; 10_000_000.times { s += "abc".hash }; p s; p Time.now - t ' --release`
963540000000
00:00:00.0001570

on pr/4675
`./bin/crystal eval 't = Time.now; s = 0_i64; 10_000_000.times { s += "abc".hash }; p s; p Time.now - t ' --release`
875314250000000
00:00:00.1288380


is it ok?

funny-falcon commented:
@kostya, yes, cause previous version calculated hash at compile time.
Whole point of this PR is to make hash unpredictable.
```
$ crystal eval 't = Time.now; s = 0_i64; ("0000001".."9999999").each {|v| s += v.hash }; p s; p Time.now - t ' --release
4911972890384
00:00:03.8365012
$ bin/crystal eval 't = Time.now; s = 0_i64; ("0000001".."9999999").each {|v| s += v.hash }; p s; p Time.now - t ' --release
Using compiled compiler at `.build/crystal'
21481679739237554
00:00:03.1407030
```

akzhan commented:
@funny-falcon just replace skip_file but do not include this change into PR.

`skip_file` fix should be subject of another PR (may be yours too).

akzhan commented:
Just to remember that it's solves #3932.

Sija commented on src/big/big_rational.cr:
```diff
@@ -41,6 +41,22 @@ struct BigRational < Number
+  # Creates a exact representation of float as rational.
+  #
+  # It ensures that `BigRational.new(f) == f`
+  # It relies on fact, that mantisa is at most 53 bits
```
mmm, mantisa!

![image](https://user-images.githubusercontent.com/988/27936843-42099426-62b3-11e7-8213-c00c8076c9d6.png)

akzhan commented on src/big/big_rational.cr:
```diff
@@ -64,8 +80,12 @@ struct BigRational < Number
+
   def <=>(other : Float)
-    self.to_f <=> other
+    BigFloat.new(self) <=> BigFloat.new(other)
```
```crystal
  def <=>(other : Float)
    to_big_f <=> other.to_big_f
  end
```

will be optimized for `other : BigFloat` case.

akzhan commented on src/float.cr:
```diff
@@ -148,13 +149,38 @@ struct Float32
+        # matches and float is IEEE754.
+        unsafe_int = unsafe_as(Int32)
+        exp = (((unsafe_int >> 23) & 0xff) - 127)
+        mantisa = unsafe_int & ((1 << 23) - 1)
```
`mantisa` here and below not yet fixed :)

akzhan commented on src/big/big_rational.cr:
```diff
@@ -139,8 +159,19 @@ struct BigRational < Number
+      floor = div.tdiv(hash_modulus)
+      rem -= floor * hash_modulus
+    end
+    BigFloat.new(rem).hash_normalize
```
It looks like BigFloat.new(self) everywhere should be replaced with `to_big_f` with copy of `to_big_f` too. 

Code will be more readable and to strange require needed.

funny-falcon commented:
Something prevents it from being merged?

akzhan commented:
@funny-falcon It's time to release 0.23.1+ with build fixes afaik.

Our pull requests anyway should be scheduled to 0.24.

RX14 commented:
It's quite a large PR so will take time to review, i'm working through a full review currently.

RX14 commented on src/big/big_float.cr:
```diff
@@ -17,6 +18,23 @@ struct BigFloat < Float
     LibGMP.mpf_init_set_str(out @mpf, str, 10)
   end
 
+  def initialize(num : BigInt)
```
Shouldn't this be in a separate PR?

akzhan replied:
It's part of `Object.hash` condition (that's wrong on current Crystal 0.23), read the [doc](https://crystal-lang.org/api/0.23.0/Object.html#hash-instance-method):

> This method must have the property that a == b implies a.hash == b.hash.

It's part of "change computation of hash value". Splitting of normalization and computation looks bad for me.

akzhan replied:
@RX14 btw, these constructors also part of another #4653 pull request. They are cross in some areas.

funny-falcon replied:
@akzhan, sorry, either you do it wrong way, or Crystal does it wrong way.
In fact, your failed tests were just wrong, cause you assumed `BigFloat.new(some_string).to_s == some_string`, but it is wrong! BigFloat - is not decimal float, but binary float.

funny-falcon replied:
Excuse me, I was wrong about your tests :-(

funny-falcon replied:
Either way, I change constructor to do the right thing instead of removing constructor.

RX14 commented on src/float.cr:
```diff
@@ -148,13 +149,38 @@ struct Float32
+
+  def hash_normalize
+    float_normalize_wrap do
+      {% if FLOAT32_IS_IEEE754 %}
```
Are you sure this works as intended? The constant's expression looks too complicated to be usable by macros. Regardless, this would still break when cross-compiling. This check must be done at runtime and cached

funny-falcon replied:
I could revert it back to `flag?(:i686) || flag?(:x86_64) || flag?(:arm) || flag?(:aarch64)` - that will be reliable for cross-compilation.

I don't want to leave it for runtime check cause of performance.
If performance doesn't matter, then this bitwise trick isn't needed at all.
But I strongly think, performance matters.

akzhan replied:
Anyway `--cross-compile` should work under `FLOAT32_IS_IEEE754z conditions.

Another case - that's cross compile option is wrong at all and must be reimplemented. This is not subject of this PR.

RX14 commented on src/float.cr:
```diff
@@ -206,11 +232,41 @@ struct Float64
+
+  def hash_normalize
+    float_normalize_wrap do
+      {% if FLOAT64_IS_IEEE754 %}
```
ditto

RX14 commented on src/hash/hasher.cr:
````diff
@@ -0,0 +1,225 @@
+# end
+# ```
+
+class Hash
````
I'm not sure this should be underneath the `Hash` class.

akzhan replied:
I'm unsure too. But `Hasher` looks like too generic to me. Feel free to propose another one.

RX14 commented on src/hash/hasher.cr:
```diff
@@ -0,0 +1,225 @@
+  #
+  # To protect against HashDos, it is seeded with secure random, and have
+  # permutation that hard to forge without knowing seed and seeing hash digest.
+  struct Hasher
```
I have no idea how to review this class: it doesn't seem to have an algorithm named so I can't look at whether it's an acceptable algorithm or check the implementation.

akzhan replied:
Nice catch. As far as I seen it's https://github.com/funny-falcon/funny_hash#funny_hash - 
but should be documented.

RX14 replied:
Has funny-hash been reviewed by people who are experts in hashing? This might not be security sensitive but I don't think we should use an algorithm that hasn't been used before or chosen by an expert in the field. I don't even know the criteria which are important when choosing a hash for this purpose let alone whether it meets those criteria.

akzhan replied:
@RX14 I'm not expert but it looks very like murmur hash. We will free to change algorithm.

funny-falcon replied:
I'm not "well known expert", and I cannot proof anything with mathematics.

It is not "funny_hash", cause llvm sucks :-( . I made special hash algorithm for Crystal to workaround all llvm optimizer issues ;-( Initially I wanted to put `lucky777` (also mine algorithm, but heavilly inspired by well known Marvin32), but llvm failed to produce rot instructions when optimized that.

I've made this hash with all my experience to avoid trivial seed independent collisions. So it is *hard* to break without knowing seed and/or result, but *not impossible*.
I tried to make it hard enough so HashDoS will be as difficult as pure DoS, so attacker will choose to perform DoS directly than spend several days to prepare HashDoS.

akzhan replied:
`lucky777` was used by picolisp afais :)

funny-falcon replied:
@akzhan , no, there is a man, who is picolisp fan, and he implemented it just for fun.

RX14 replied:
@funny-falcon I'm somewhat sceptical that LLVM is the issue here, have you tried compiling your funny_hash implementation with clang and comparing to GCC?

funny-falcon replied:
Yes, I did. It was hard to force clang to do the same as gcc.

I don't remember, if I tried funny_hash for crystal. I certainly tried lucky777, cause it is faster than funny_hash on gcc. But after crystal's aggressive inlining, llvm failed to combine shift+or into rotates. llvm really sucks on this optimisation. Why the hell they didn't include rotate bitcode, so frontend may produce it before llvm invoked at all?

I can retry funny_hash for Crystal. funny_hash is safer, but it almost certainly will be slower than already suggested hash. And I think suggested hash is safe enough.

RX14 replied:
Well there's probably a llvm rot intrinsic you could bind.

akzhan commented on spec/std/big/big_rational_spec.cr:
```diff
@@ -149,7 +149,7 @@ describe BigRational do
     b = br(10, 3)
     hash = b.hash
-    hash.should eq(b.to_f64.hash)
+    hash.should eq(b.to_big_f.hash)
```
https://play.crystal-lang.org/#/r/2cnl

10/3 hash calculated using BigFloat precision.

funny-falcon replied:
Float64 has only 53bit meaningful mantisa, but for hash calculation I use 61bit Mersen prime.
That is why conversion to Float64 looses precision.

As an alternative, it could stick with 31bit Mersen prime. Then here could be `to_f64`.
But then Int32, UInt32, Float64 hashing will have to perform modulo operation.

I believe, hashing Int32, UInt32 is much more important, than hashing BigRational and BigFloat.
That is why I decided to use 61bit Mersen Prime, and fallback to `to_big_f` at this place, and to calculate BigFloat hash by algorithm, instead of using `to_f64`.

It is just tradeoff. If you will decide, that BigRational and BigFloat is more important, than Float64 and Int32, I will change it back.

akzhan commented on src/float/hash_normalize.cr:
````diff
@@ -0,0 +1,78 @@
+  #                              -1.5009765625_f64.unsafe_as(UInt64) == 0xbff8040000000000_u64
+  # ```
+  # But to simplify cross-compilation, just hardcode it.
+  private FLOAT64_IS_IEEE754 = {{ flag?(:i686) || flag?(:x86_64) || flag?(:arm) || flag?(aarch64) }}
````
:aarch64 ?

funny-falcon replied:
yes, thanks!

funny-falcon commented:
I believe, fail on travis-ci is not my fault. Am I wrong?

Sija commented on src/float/hash_normalize.cr:
````diff
@@ -0,0 +1,78 @@
+  #                              -1.5009765625_f32.unsafe_as(UInt32) == 0xbfc02000_u32
+  # ```
+  # But to simplify cross-compilation, just hardcode it.
+  private FLOAT32_IS_IEEE754 = {{ flag?(:i686) || flag?(:x86_64) || flag?(:arm) || flag?(aarch64) }}
````
`aarch64` -> `:aarch64`

funny-falcon replied:
done

akzhan commented:
Just for accuracy: Hashing is not at all a high science, its task is only to ensure a good distribution with high performance.

This request solves three problems,
- Protection against HashDoS attacks, so that it would be cheaper to conduct a normal DoS against the application, as Yura noted.
- Ensuring equality of hashes with equality of values for numbers, as the language itself requires.
- Convenient API for the possibility in the future to easily change or extend the hashing algorithm taking into account modern requirements for hashing.

Sija commented on src/big/big_float.cr:
```diff
@@ -30,6 +30,24 @@ struct BigFloat < Float
       LibGMP.mpf_set(self, num)
+    when Int8, Int16, Int32
+      LibGMP.mpf_init_set_si(out @mpf, num)
+    when Int8, Int16, UInt32
```
did you meant `UInt8, UInt16, UInt32`?

funny-falcon replied:
yes, you are right.

Sija commented on src/big/big_float.cr:
```diff
@@ -18,7 +18,39 @@ struct BigFloat < Float
+    when BigFloat
+      LibGMP.mpf_init(out @mpf)
+      LibGMP.mpf_set(self, num)
+    when Int8, Int16, Int32
```
did you meant `UInt8, UInt16, UInt32`?

Sija commented on src/big/big_float.cr:
```diff
@@ -18,7 +18,39 @@ struct BigFloat < Float
+      LibGMP.mpf_set(self, num)
+    when Int8, Int16, Int32
+      LibGMP.mpf_init_set_si(out @mpf, num)
+    when Int8, Int16, UInt32
```
did you meant `UInt8, UInt16, UInt32`?

funny-falcon commented:
It was hard :-(
LibGMP::Long and LibGMP::ULong is Int32 and UInt32 on 32bit platform, so there is a lot of bugs in big_int.cr/big_float.cr . I fixed some in big_int.cr, but I think, there more.

akzhan commented:
Yep, I'm already annoyed by strange aliasing of basic types (like SizeT/ULong does).

And yes, Crystal stdlib has bad assumptions through...

Sija commented on src/number/hash_normalize.cr:
```diff
@@ -41,24 +41,14 @@ module Number::HashNormalize
+
+  private HASH_BITS = 61
+  private def hash_bits
+    61
```
Shouldn't this return `HASH_BITS`?

funny-falcon replied:
I add this method to not refer any constant.
See my comment above in this discussion about weird way Crystal handles constants.

Sija replied:
How about adding [`@[AlwaysInline]`](https://github.com/crystal-lang/crystal-book/blob/master/syntax_and_semantics/attributes.md#alwaysinline) attribute?

funny-falcon replied:
AlwaysInline will not hurt. But looks like it is not needed for such simple method in release build.
For debug build?

funny-falcon commented:
Friends, help me how to pass travis-ci? crystal could not fork on compiling all_spec cause it is too fat.
Why it forks at all?

akzhan commented:
You aren't alone. @RX14 already on it.

See https://github.com/crystal-lang/crystal/pull/4707#issuecomment-315196808 for details.

funny-falcon commented:
Oh, I thought I made crystal too irregular :-) Last several commits were mostly cause of that.
I'd better will not revert them, cause code is a bit simpler now (it is less portable in theory, but in practice no bad things introduced).

akzhan commented:
Anyway b32be7d4073ba198ee9c87dae656902a39ebad9f looks wrong, it should be kept as comment, for example.

funny-falcon commented:
there is still comments: https://github.com/funny-falcon/crystal/blob/38ef903d3638971e2984510e0186fe653893bb78/src/float.cr#L245-L247
I doubt crystal will support anything with non-IEEE754 floats at near future (cause I think LLVM doesn't support such architectures; I could be mistaken).

akzhan commented:
@funny-falcon So you may to add this definition as comment. Just for next generations.

funny-falcon commented:
> Looks like we need to revert hashing algorithm to custom one.

@akzhan , why do you say it? I don't get the point.

akzhan commented:
custom (lucky777 inspired) algorithm was last that passed the spec

funny-falcon commented:
Travis-ci first failed on old version.
Looks like travis-ci was changed: memory were restricted.

funny-falcon commented:
Would you mind I reopen new PR with squashed commits?
Should I restore "float/hash_normalize.cr" and "FLOAT32_IS_IEEE754" ?

RX14 commented:
@funny-falcon Squashed, topic-based commits would be great, but you don't need a new PR for it.

bararchy commented:
@funny-falcon @RX14 
Is this getting merged ? or does it need more changes ?

funny-falcon commented:
I believe, it could be merged.

On the other hand, it could be split into #4653, #4581 (should be reopened), and hashing itself.
@akzhan likes this idea, but I have no enough time to do the split.

RX14 commented:
Can you move the stringpool stuff to a seperate PR? Also my complaint about the algorithm used still stands.

funny-falcon commented:
> Also my complaint about the algorithm used still stands.

What is complaint about?
Yes, I have no authority to proof anything. Though I strongly believe it is just good enough.
And it is certainly better than current way.
And it could be easily replaced in a future if any flaw will be discovered. Hashing infrastructure introduced is mostly algorithm agnostic, so if someone will prove this algorithm is not secure enough, then it could be easily replaced.

> Can you move the stringpool stuff to a seperate PR?

If you want so. But it is strange to make whole thing safer, and leave stringpool (that is used in json decoding) unsafe.

straight-shoota commented:
> if someone will prove this algorithm is not secure enough

It is usually the other way around: You prove that an algorithm is reasonably secure and then it may be considered to be employed.
Though you're right that this is better than the status quo, so it would be an improvement anyway. But it would probably be worth to directly use a proven algorithm, even if this means it takes longer to be merged.

RX14 commented:
As @straight-shoota said, the burden of proof is on you to prove that the algorithm is secure *in a peer-reviewed paper*, not on us to prove it insecure when we know nothing about the problem domain.

funny-falcon commented:
look, there is only one "proven" algorithm: siphash. It is slow. But it is "cryptographic", so there are papers.

No one seriously (ie with scientific paper) will bother proving non-cryptographic algorithm.

If you want, you may use SipHash by default. I wash my hands.

RX14 commented:
To summarise a conversation on IRC: all we need is a second opinion on this algorithm, just like all PRs need 2 core team members to sign off on them. I'm perfectly happy to merge this once we have a second opinion. If you would rather get the ball rolling now, another option would be to merge siphash13 first and this algorithm later. I don't mind either way, but I think this breaking change should be done in the next few months.

akzhan commented:
just to not forget:


Sokolov Yura @funny-falcon aug. 07 17:54

> 
> Yes. Current code in PR is mostly funny_hash, and I've tested funny_hash with SMHasher.
> (I've tested previous variants of code (a bit faster hash function specially crafted for Crystal) in SMHasher as well).
> 
> SMHasher is just statistical analyzer, it tells "your function does not obviously suck". It is not proof of safety. Murmur3a passes SMHasher as well.
> 
> Safety comes from construction: Murmur3a fails cause man found simple and cheap way to generate collisions regardless of seed. There is a way to construct just 1 bit difference between blocks, and then cancel it. And attacker doesn't need to know anything about seed or result hashsum to construct such collision.
> funny_hash constructed in a way, every block bit affects at least several state bits, and there is no obvious seed independent way to cancel all difference bits in a state with high enough probability. And every not-canceled bit increases state avalanche significantly.
> 
> I think, there is a way to construct collisions, if hashsum is exposed to attacker, though. That is a border between "cryptographic" and "non-cryptographic" function: if knowing something about hashsum helps attacker, it is not cryptographic function.
> (A bit more complex finalizer significantly increase effort for such collision construction. But given, "exposed hashsum" is not intended use-case, I prefer faster variant for falcon_hash.)
> 

Sokolov Yura @funny-falcon aug. 07 18:07

> 
> And I should correct myself: SipHash13 is slower for short strings (cause its finalizer is heavy weight), and it is slower on 32bit platform (cause it uses 64bit arithmetic). funny32_hash is faster on short strings on any platform, and on 32bit platform for any string.
> 
> There is funny64_hash that uses 64bit integers that is always faster than SipHash13, but it is slower on 32bit platform than funny32_hash. I didn't include it into PR to simplify code.
>

ysbaddaden commented:
The whole topic is about mitigating hash flooding. This pull request does a better job, but I'm warry of the chosen funny-hash algorithm, which no peer reviewed formally. I'm incapacable to tell whether the algorithm is good, correct or just plain bad. I don't have the technical or mathematical knowledge to achieve this.

No hard feelings, but we should rely on a peer-reviewed algorithm. SipHash was created to fix this specific issue: hash flooding attacks; siphash-2-4 was peer reviewed, and acknowledged to be cryptography secure.

Since the hashes are never disclosed to an attacker, we don't need the cryptographically secure version (siphash-2-4) and it seems acceptable to mitigate safety for speed, and rely on siphash-1-3 or even halfsiphash-1-3 as per this thread on LKML: https://lkml.org/lkml/2016/12/19/492

I'd prefer this pull request to use siphash-1-3 on 64-bit systems, and halfsiphash-1-3 on 32-bit systems.

funny-falcon commented:
As I said before, this PR will not use SipHash, but I don't mind if most of code will be copied to other PR.
Regards.

RX14 commented:
OK then shall we close this?

akzhan commented:
@RX14 You are wrong here. We can pull all changes now (just because it's fine, very attractive and done using Crystal specifics). 

And we can to replace hash algorithm by another PR (but I don't know why, it's passed SMHasher tests).

RX14 commented:
If we're just going to replace the hash algorithm anyway why not squash these commits and fixup the new algorithm into them. 

The simple fact is that using an uncommon hashing algorithm is a potential risk, and I (we) don't like taking risks around security.

funny-falcon commented:
It is funny: before I noticed and before this PR there was just dumb "h =  h*31 + c", and doesn't seems anybody care (about hash-flooding).

I opened the problem, cause I care. I suggested solution I've invented to solve the problem. But now no one trust my solution, cause "it wasn't peer reviewed" and "has potential risk" and "you have no authority".

Fair.

I'm tired to react on this discussion. Please, copy code, change algorithm, do what you want in other PR.

akzhan commented:
Jean-Philippe Aumasson <jeanphilippe.aumasson@gmail.com> wrote:

> Hi!
>
> had a quick look, wouldn't be surprised if it were vulnerable to some form of differential cryptanalysis. Also, looks like weak seeds would make (near-)collisions easier to find. For example, iiuc, with the all-zero seed and two all-zero byte arrays, the state will remain zero until the last step where the array length is ingested.
> 
> This hash may or may not be secure enough, guess it would need more cryptanalysis.
> 
> Hope it helps,
> 
> JP

funny-falcon commented:
@akzhan, thank you for publishing Jean-Philippe's review. It contains nothing new for me though:

- differential analyze could be performed if hashsum is exposed to attacker, and it is not the intended use-case for hash function,
- zero seed with zero array is so obvious case so no need to be cryptographer to recognize it. But given seed is taken from strong source, probability of such seed is 2^-64.

My function is not cryptographic, and I never said that. I'm saying: it could make hash-flood as hard as pure DoS if strong per-process seed is used and hashsum of any value is not exposed to attacker.

Still, please, don't write to this PR any more. It is hard to me not to answer.

funny-falcon commented:
@akzhan confirmed that he didn't mentioned "protection against hash-flood" in letter to Jean-Philippe, but formulated as "its own algorithm alternative to SipHash".

straight-shoota commented:
@funny-falcon I am sure you did a great effort to create a good hashing algorithm for Crystal. Thank you for that! I guess no one doubts that. But there is also no one with sufficient experience to confirm it.
So it's understandable that the core team won't take the risk just based on your assessment and rather prefers a proven alternative even if your's might be better in some regards.

You sound disappointed, but I hope there are no hurt feelings.
On retrospect I believe this could have been much better with proper communication. Had you proposed your algorithm first, it could haven been clearly stated, that it won't be merged unless properly proven, before you implemented it and created the PR (#4675). And I might also have been better to separate the abstractions for generating hashs and updating the hashing algorithm into separate PRs. Most of the discussion was about the integration of `Hasher` not the actual algorithm. And I think this is really a great improvement.

You're right that the current implementation is far worse in all regards. But it was only meant as a interim solution, now the goal is to have something that can ship with a stable 1.0 release. There is little sense in replacing an algorithm that is (obviously) not fit for this for one that is (until proven otherwise) also not considered usable for this.

funny-falcon commented:
@straight-shooter , as I've said several times before:

- I don't mind if one change algorithm *after* merging this PR,
- I don't mind if one copy whole this PR (or most of it) into *new PR* and change algorithm to whatever he/she wants.

I just don't want do it by myself in *my* PR.

My PR contains my vision on:

1. on high level infrastructure for hashing,
2. on low-level hash function.

Single algorithm change I can accept in my PR is returning to `h = h*31 + c` , ie return low-level function to status-quo, and leave only high-level infrastructure changes in PR.

RX14 commented:
@funny-falcon I think that last option was what @straight-shoota suggested. Perform the Hasher abstraction in this PR then replace the terrible Hasher impl we would have with another hash algorithm (99% sure we'd choose SipHash-1-3 and halfSipHash-1-3 on 32bit)

funny-falcon commented:
@RX14 ok, I'll do it.

ysbaddaden commented:
Thank you :heart:

funny-falcon commented:
I "reverted" algorithm to "h = h*31 +v". I leaved a bit of finalizer to make function sane with floats though.
Also I leave seeding with random value as an example for future implementations (though it doesn't matters much with this algo).

I removed commit for StringPool. Just don't forget about it later, please.

RX14 commented:
@funny-falcon I'm sorry to mislead you but I've actually changed my mind about the stringpool commit. It belongs in this PR just as a seperate commit.

funny-falcon commented:
@RX14 no problem, I put StringPool commit back.

RX14 commented:
@funny-falcon Any data on stringpool performance before/after ?

funny-falcon commented:
No, I didn't bench it. But given I simplified types and removes couple of indirections (`Slice(String) + Slice(Uint32)` instead of `Array(Array(String)?)`) and made hashsum comparison before string comparison, I doubt performance could degrade. At least it should be not slower.

funny-falcon commented:
looks like I was mistaken about StringPool performance a bit: 
- searching is always faster,
- but creation is slower due to allocation of larger arrays.

RX14 commented:
It does speed up parsing large JSON data noticably (by about 3.5% on this specific dataset).

![](https://my.mixtape.moe/ktufww.png)

akzhan commented:
```
➜  crystal git:(9640f8028) ✗ bin/crystal samples/bench_str_pool.cr --release --no-debug
Using compiled compiler at `.build/crystal'
insert new string pool    3.1  ( 322.2ms) (± 1.99%)       fastest
insert old string pool   2.87  (348.24ms) (± 2.93%)  1.08× slower
get new string pool   3.06  (327.18ms) (± 2.99%)       fastest
get old string pool   2.88  (347.32ms) (± 3.33%)  1.06× slower
```

source code:

```crystal
require "benchmark"
require "string_pool"
require "old_string_pool"

Benchmark.ips do |x|
  x.report("insert new string pool") { test_with(StringPool.new) }
  x.report("insert old string pool") { test_with(OldStringPool.new) }
end

Benchmark.ips do |x|
  x.report("get new string pool") { test_with(StringPool.new) }
  x.report("get old string pool") { test_with(OldStringPool.new) }
end

def test_with(pool)
  (0..1_000_000).each do |num|
    pool.get("hello #{num}")
  end
end
```

funny-falcon commented:
Ah, I see, in my test creation were slower cause new version did more `rehash` calls (it increased table x2, but old version x5). I changed it to increase x4, and now it is always not slower.

And I fixed formatting hash/hasher.cr
(pushed branch with force)

akzhan commented:
updated to latest commit (5d0c15ae4) results:

```
➜  crystal git:(5d0c15ae4) ✗ bin/crystal samples/bench_str_pool.cr --release --no-debug
Using compiled compiler at `.build/crystal'
insert new string pool   3.74  (267.47ms) (± 7.46%)       fastest
insert old string pool   2.74  (364.52ms) (± 1.03%)  1.36× slower
get new string pool   3.84  (260.15ms) (± 1.21%)       fastest
get old string pool   2.69  (372.37ms) (± 3.37%)  1.43× slower
```

RX14 commented:
Whereas my samples show this:

![](https://my.mixtape.moe/trklrj.png)

Sija commented on src/hash/hasher.cr:
```diff
@@ -0,0 +1,188 @@
+      @a = @a * 31 + v
+    end
+
+    protected def permute(v : UInt32)
```
Why not combine two of those into `protected def permute(v : UInt8 | UInt32)`?

funny-falcon replied:
Cause it is "example template" for future algorithm.

funny-falcon commented:
@RX14 Could you test new version (i pushed just now) ? I return increase x2, but speedup rehashing itself.

RX14 commented:
I redid the "patch applied" results and my graph seems to now be mostly noise. I think I need a dedicated computer to run these tests on but setting up a crystal dev env is a pain in the arse on the debian-based distributions which cloud providers like to use. So please, consider my results as bogus:

![](https://my.mixtape.moe/hxmkqn.png)

Sija commented on src/string_pool.cr:
```diff
@@ -26,9 +26,10 @@ class StringPool
+    @hashes = Pointer(Hash::Hasher::Value).malloc(8, 0_u32)
+    @values = Pointer(String).malloc(8, "")
     @size = 0
+    @capa = 8
```
`@capa` => `@capacity`

funny-falcon replied:
done

hugoabonizio commented on src/string_pool.cr:
```diff
@@ -149,20 +149,20 @@ class StringPool
     end
 
-    old_capa = @capa
+    old_capa = @capacity
```
s/old_capa/old_capacity :smile:

funny-falcon replied:
done

Sija commented on src/big/big_float.cr:
```diff
@@ -35,8 +67,19 @@ struct BigFloat < Float
-  def hash
-    to_f64.hash
+  def hash_normalize
+    # more exact version of `remainder(hash_modulus).to_f.hash_normalize
```
Missing trailing backquote.

Sija commented on src/number/hash_normalize.cr:
```diff
@@ -0,0 +1,105 @@
+  end
+
+  private def hash_infinity
+    int_to_hashnorm 314159
```
Shouldn't that be a constant?

funny-falcon replied:
Done: https://github.com/crystal-lang/crystal/issues/4884

RX14 commented on src/big/big_float.cr:
```diff
@@ -18,7 +18,39 @@ struct BigFloat < Float
+    # remove it when separate BigFloat.new(BigInt) will pass the spec
+    case num
+    when BigInt
+      LibGMP.mpf_init(out @mpf)
```
I forgot this PR depended on https://github.com/crystal-lang/crystal/pull/4653. We should really get that PR merged so that this PR can be rebased on top of it.

funny-falcon replied:
This PR doesn't depend on #4653 , but they have a lot of code in common. So, if one merged, other should be rebased.

RX14 replied:
Well the changes in #4653 came from this PR and have already been split out and changed a bit. If it's already been split out then it makes a lot of sense to review that first so we can ensure that all the required changes make it in and we're not splitting the review effort over 2 PRs.

funny-falcon replied:
Fair. I don't mind if #4653 will be merged first.

akzhan replied:
@RX14  feel free to merge this PR firstly. 
There a lot of common code, because we have no assumptions on its pulling order.

I will rebase my PR anyway.

funny-falcon commented on src/number/hash_normalize.cr:
```diff
@@ -0,0 +1,105 @@
+  end
+
+  private def hash_infinity
+    int_to_hashnorm 314159
```
Crystal currently produces not very efficient code with constants, so I avoided them.

funny-falcon replied:
But call to int_to_hashnorm is artefact of old versions. It could be replaced with literal u64.

RX14 replied:
Please make an issue about the compiler constant problem. I think that literal values and structs shouldn't be initialized on demand.

Fryguy commented:
Does this also fix #2817 if merged?

funny-falcon commented:
@Fryguy it is half of equation. Second half is appropriate hash algorithm. My algorithm were rejected, so I removed it from this PR. See discussion above.

Most probably, some algorithm will be merged soon after merging this PR.

RX14 commented:
Ok, we're good to rebase this on `master` and another quick review and we're good to merge.

funny-falcon commented:
Ok, will do :-)

akzhan commented:
@funny-falcon will be nice to use benefits provided by #4904.

funny-falcon commented:
@akzhan , certainly :-)

funny-falcon commented:
Ok, rebased, changed methods to constants. Added couple of 32bit fixes to big_int.

Sija commented on src/number/hash_normalize.cr:
```diff
@@ -0,0 +1,92 @@
+  # P is of the form 2**n-1, reduce(2**e) is 2**(e mod n), and multiplication
+  # by 2**(e mod n) modulo 2**n-1 just amounts to a rotation of bits.
+
+  # XXX: private methods used instead of constants cause current crystal
```
Isn't this comment obsolete after #4904?

funny-falcon replied:
Yeah, forgot about.

Sija commented on src/string_pool.cr:
```diff
@@ -26,8 +26,10 @@ class StringPool
   # Creates a new empty string pool.
   def initialize
-    @buckets = Array(Array(String)?).new(11, nil)
+    @hashes = Pointer(Hash::Hasher::Value).malloc(8, 0_u32)
```
This should use `@capacity` instead of fixed `8`.

funny-falcon replied:
will do

Sija commented on src/string_pool.cr:
```diff
@@ -26,8 +26,10 @@ class StringPool
   def initialize
-    @buckets = Array(Array(String)?).new(11, nil)
+    @hashes = Pointer(Hash::Hasher::Value).malloc(8, 0_u32)
+    @values = Pointer(String).malloc(8, "")
```
ditto

funny-falcon replied:
ditto

akzhan commented:
@funny-falcon  Thanks again for 32bit plaform fixes (BigInts). 

They are robust, and LGTM.

I was cherry-picked 'em to minor #4922 pull request, but feel free to merge 'em in any order.

akzhan commented:
@funny-falcon please rebase your pull request (32bit BigInt patch applied to master).

@RX14 looks like it's ready to merge.

funny-falcon commented:
I've rebased. Hope this saga will end :-)

RX14 commented on src/hash/hasher.cr:
```diff
@@ -0,0 +1,188 @@
+    # Type for Hash::Hasher#digest
+    alias Value = UInt32
+
+    @@seed = uninitialized StaticArray(UInt32, 1)
```
Why not just make this a random U32? If we use a constant with initializer (`SEED = Random::System.rand(UInt32::MIN..UInt32::MAX)`), shouldn't the constant be initialized on-demand, avoiding the need for special "unseeded" methods below?

RX14 commented on src/hash/hasher.cr:
```diff
@@ -0,0 +1,188 @@
+    buf = pointerof(@@seed).as(Pointer(UInt8))
+    Crystal::System::Random.random_bytes(buf.to_slice(sizeof(typeof(@@seed))))
+
+    protected getter a : UInt32 = 0_u32
```
Would be nice if this was renamed to something a bit more descriptive.

RX14 commented on src/hash/hasher.cr:
```diff
@@ -0,0 +1,188 @@
+    end
+
+    # Calculate hashsum for value
+    def self.hashit(value) : Value
```
I think i'd just name this `hash`.

RX14 commented on src/hash/hasher.cr:
```diff
@@ -0,0 +1,188 @@
+    # Mix nil to state
+    def <<(v : Nil) : Nil
+      permute_nil()
+      nil
```
You can leave these trailing `nil`s off if we have the ` : Nil` restriction, iirc.

RX14 commented on src/hash/hasher.cr:
```diff
@@ -0,0 +1,188 @@
+    end
+
+    # Mix nil to state
+    def <<(v : Nil) : Nil
```
The only function of these methods seem to be calling `permute`. Why not move the permute body into this function?

RX14 commented on src/hash/hasher.cr:
```diff
@@ -0,0 +1,188 @@
+    protected def permute_nil
+      # LFSR
+      mx = (@a.to_i32 >> 31).to_u32 & 0xa8888eef_u32
+      @a = (@a << 1) ^ mx
```
Isn't the current implementation of this essentially `@a *= 31` not a LFSR?

RX14 commented on src/http/headers.cr:
```diff
@@ -9,13 +9,12 @@ struct HTTP::Headers
-        c = normalize_byte(c)
-        h = 31 * h + c
+    def hash(hasher)
+      hasher.raw(bytesize.to_u32)
```
Why raw here? I think that the correct way to use hasher should *always* be `<<` without anyone having to think about `raw` except people implementing `hash(hasher)` on numbers. Shouldn't `raw` be `:nodoc:`?

RX14 commented on src/string_pool.cr:
```diff
@@ -70,26 +72,45 @@ class StringPool
-      if entry
-        return entry
+    mask = (@capacity - 1).to_u32
+    index, d = hash & mask, 1
```
please split this into two assignments for readability.

funny-falcon commented:
Chris, most of your last remarks are meaningless, given it is just template for stronger algorithm you will have to implement soon after merging this.

 Renaming `hashit` to `hash` is quite ridiculous: having same name+arity combinatiom for different semantic actions is error prone.

I will not waste my time on this PR anymore.

funny-falcon commented:
And why do you think I use class variable instead of constant? To avoid "initialization on demand" inlined into every call to `hash`.

RX14 commented:
@funny-falcon seed is only used when initializing the hasher, so it's relatively off the hot path compared to say the inner loop of a hash function. Furthermore it's random so it doesn't help at all with further optimizations.

Where is there another `self.hash` method on `Hasher`?

I think that i've made some relatively minor requests for a cleanup which are relevant both to this algorithm *and* to the implemntation of any future algorithm. I see no reason why siphash-1-2 can't be implemented in `<<` instead of `permute`.

funny-falcon commented:
What if hot loop calls `val.hash` on integer, and `val.hash` it is not inlined, so SEED initialization could not be moved out of loop? What about code bloat, cause SEED initialization inlined in many places.

Isn't `Hasher` is also `Object` that already has `Object#hash(h)`?

Will you inline siphash implementation into every `<<`?

`raw` is needed cause there is difference between hashing int as number that needs normalization, and as opaque part of opaque structure, where normalisation is certainly not needed.

RX14 commented:
@funny-falcon 

I think it's likely to be a very minor performance hi, *more* than worth it for the advantage of removing the complications of having to deal with initialization order and the `unseeded` method. @asterite would likely have more knowledge of the performance of class vars vs constants here.

We're talking about a class method, `self.hash` which is entirely seperate from `#hash` on the instance.

I'm not an expert in siphash, but wouldn't you would implement siphash in terms of one input length and the other `<<` overloads would call the `<<` overload which implements siphash.

Thanks, I hadn't thought of that. It appears the only differece (currently) between `raw` and `<<` is for 64-bit signed and unsigned integers, would that change for example with siphash? Why can't we just mix in the higher then lower bits always and avoid this?

funny-falcon commented:
You arguments are out of my sense of sanity.
Perhaps you are right, but I cann't get it.
I will not change this PR any more.

asterite commented:
Class vars and non-simple constants are always lazily initialized because you never know when one constant will need the value of another constant. It's a way to avoid depending on the order of initialization.

Probably seed can be moved to a constant (that will always be lazily initialized, and checked on every access, but in any case I don't think this should affect performance much)

funny-falcon commented:
Friends, I don't want to discus it any more. You may change anything after merging. Instead you pushing on me in this small details. I think code will be worse if I apply your last suggestions. Therefore I will not apply them.

I'm too tired of this PR. If you are not going to merge it in its current shape, I will close this PR and remove code after tomorrow.

asterite commented:
@funny-falcon You are right. Open source can be a PITA, I feel the same every day.

I will take care of this issue, don't worry. Thank you for this contribution! I will probably copy some pieces of code, if that's OK with you.

funny-falcon commented:
@asterite , thank you.
Yes, you (and anybody else) can copy code from this PR as much as you want.

funny-falcon commented:
s/can/may/

faustinoaq commented:
@funny-falcon Thanks for your time invested here, Sometimes Open Source software could be very hard to manage.

I need to quote this:

> Ary: ... The community part is a whole world of its own, tackling issues and PRs, and I'd say it's even more challenging (but also rewarding!) than developing the language itself :-)

faustinoaq commented:
> I'm too tired of this PR 

I understand you, This PR is the most commented until now :sweat_smile:

I think 216 comments in this PR is a record for this project.

![screenshot_20170910_172347](https://user-images.githubusercontent.com/3067335/30253530-cfb270a0-964c-11e7-882c-65485c4ca661.png)

stay tuned of #4946

akzhan commented:
For now all great parts of #4675 now merged into `master` branch in some way.

Thanks again to all of us.
